OpenRA.Mods.Common/AI/HackyAI.cs

```diff
@@ -984,7 +987,11 @@ void BuildUnit(string category, bool buildRandom)
 				ChooseRandomUnitToBuild(queue) :
 				ChooseUnitToBuild(queue);
 
-			if (unit != null && Info.UnitsToBuild.Any(u => u.Key == unit.Name))
+			if (unit != null
+				&& Info.UnitsToBuild.ContainsKey(unit.Name)
+				&& (!Info.UnitLimits.ContainsKey(unit.Name)
+					|| World.Actors.Count(a => a.Info.Name == unit.Name && a.Owner == Player) < Info.UnitLimits[unit.Name]))
```

Comment:
@penev92 suggested that HackyAI should cache the amount of it's units into a `Dictionary` pretty much, which would probably end up even more elegant - it's just massive and I didn't felt such a scopecreeping would be nice here.